### PR TITLE
Implement CRUD operations for simple to do list app in DJANGO and JS.

### PR DESCRIPTION
**Pull request by 210029073**

### CODE REVIEW

**Review by ibrahim-2k3 — APPROVED** [review]
GET request has now been added, now API maintainers will recognise that this is a GET request.

**Review by ibrahim-2k3 — COMMENTED** [review]
Do the same with the other API requests you have done so far, instead of using XMLHttpRequest object in Java, use the asynchronous fetch() API method call in JS to perform the API requests.

**Review by 210029073 — COMMENTED** [review]
Please see how to  [automatically perform method call from a js file in a html file](https://stackoverflow.com/questions/44815288/how-do-you-automatically-run-an-external-javascript-function-with-a-variable)

File: todosproject/todosapp/static/todosapp/autoload.js [hunk 1]
```diff
@@ -24,18 +24,21 @@ async function fetchData() {
...
+  return parsed_data;
 }
 
 fetchData(); //make the api request for getting todo list data
 
 /**
  * This will initialise the data in response body which is the target webpage.
  */
-function initialize() {
+async function initialize() {
```
Comment (210029073) [inline todosproject/todosapp/static/todosapp/autoload.js:37]: `initialize()` is now made as async

File: todosproject/todosapp/static/todosapp/autoload.js [hunk 2]
```diff
@@ -24,18 +24,21 @@ async function fetchData() {
...
+  parsed_data = await fetchData();
```
Comment (210029073) [inline todosproject/todosapp/static/todosapp/autoload.js:38]: This should now resolve the issue when listing the items in the webpage.

**Comment by 210029073** [comment]
This will provide documentation on how to pass json data to server via fetch() api method call. Although Flask documentation, the js part of the documentation may help in django method. https://flask.palletsprojects.com/en/2.3.x/patterns/javascript/

File: todosproject/todosapp/static/todosapp/autoload.js [hunk 3]
```diff
@@ -148,15 +148,19 @@ async function deleteItem(task) {
...
-    }
 
-    //add to session storage
-    sessionStorage.setItem("pending-task", JSON.stringify(element));
+      if (sessionStorage.getItem("pending-task") != null) {
+        sessionStorage.removeItem("pending-task");
+      }
+
+      //add to session storage
+      sessionStorage.setItem("pending-task", JSON.stringify(element));
```
Comment (210029073) [inline todosproject/todosapp/static/todosapp/autoload.js:161]: start and due dates lose their original value, causing it to be one hour backwards in GMT time.

Comment (210029073) [reply todosproject/todosapp/static/todosapp/autoload.js:161]: May need to parse the `element` iterator as a `dict` object.

Comment (210029073) [reply todosproject/todosapp/static/todosapp/autoload.js:161]: Frontend time differs with backend time.

File: todosproject/todosapp/static/todosapp/autoload.js [hunk 4]
```diff
@@ -72,7 +72,9 @@ function prepare() {
       new Date(element.duedate).toLocaleString() +
```
Comment (210029073) [inline todosproject/todosapp/static/todosapp/autoload.js:72]: Date time is not accurate due to using locale

Comment (210029073) [reply todosproject/todosapp/static/todosapp/autoload.js:72]: Should now work with `toUTCString()`

File: todosproject/todosapp/static/todosapp/autoload.js (hunk 3 above)
Comment (210029073) [reply todosproject/todosapp/static/todosapp/autoload.js:161]: image

This is the fix.

File: todosproject/todosapp/static/todosapp/autoload.js [hunk 5]
```diff
@@ -66,10 +66,10 @@ function prepare() {
       element.title +
       "</p>" +
       "<p>" +
-      new Date(element.startdate).toLocaleString() +
+      new Date(element.startdate).toUTCString() +
```
Comment (210029073) [inline todosproject/todosapp/static/todosapp/autoload.js:69]: utc time is now used here.

File: todosproject/todosapp/static/todosapp/autoload.js [hunk 6]
```diff
@@ -66,10 +66,10 @@ function prepare() {
...
       "</p>" +
       "<p>" +
-      new Date(element.duedate).toLocaleString() +
+      new Date(element.duedate).toUTCString() +
```
Comment (210029073) [inline todosproject/todosapp/static/todosapp/autoload.js:72]: same here...

File: todosproject/todosapp/static/todosapp/autoload_update.js [hunk 7]
```diff
@@ -25,16 +28,53 @@ async function initialize() {
...
+  console.log(task);
+  task.title = document.getElementById("update-task").elements[0].value;
+  task.startdate =
+    document.getElementById("update-task").elements[1].value + "Z";
+  task.duedate = document.getElementById("update-task").elements[2].value + "Z";
+  console.log("Task new");
+  console.log(task);
+
+  //retrieve csrftoken
+  part = document.cookie.split("csrftoken=");
```
Comment (210029073) [inline todosproject/todosapp/static/todosapp/autoload_update.js:48]: May require something else effective for the long-run, as the current solution may not work if there are multiple cookies.

Comment (210029073) [reply todosproject/todosapp/static/todosapp/autoload_update.js:48]: Besides splitting cookie.

Comment (210029073) [reply todosproject/todosapp/static/todosapp/autoload_update.js:48]: A sophisticated regular expression will do fine.

**Pull request merged**